### ISSUE TITLE
Batched matmul / einsum / batched tensordot

### ISSUE THREAD

**Issue by arogozhnikov** [data-apis/array-api#727]
Currently I don't see a way to implement einsum or batched matmul using the standard. There is a way with explicit loop, but that does not count.

Some ways this can be addressed:

- minimal implementation would include batched_matmul (signature `bij, bjk -> bik`). Other operations can be reduced to it with reshapes. This may include additional copies
- modification of tensordot to include batched dimensions thus should be more efficient, but AFAIK this method is not surfaced in any major framework
- include einsum in the standard

**Comment by leofang** [comment]
matmul already supports batching in the spec. In fact, batching is the [number one requirement for linalg routines](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html#design-principles), so wherever it's applicable we allow it.  

Batched tensordot is indeed like you said not implemented in any framework, most likely because there's matmul and einsum in the existing frameworks. Semantically, it's not straightforward to allow batching while keeping the same API style (I think it likely needs an extra kw-only argument to specify the batch dimensions).

I was tasked to propose einsum, as I had strong opinons after the involvement in implementing cuQuantum's counterpart, but unfortunately I lacked of time. I have a very radical version in mind (some NumPy behavior is simply bad IMHO) that has not been implemented in all but a few (cuQuanutm & opt_einsum) libraries. Feel free to share your thoughts or even draft a PR, we'll get it discussed 🙂

**Comment by arogozhnikov** [comment]
> matmul already supports batching in the spec. In fact, batching is the [number one requirement for linalg routines](https://data-apis.org/array-api/latest/extensions/linear_algebra_functions.html#design-principles), so wherever it's applicable we allow it.

Great, I've missed that

> (I think it likely needs an extra kw-only argument to specify the batch dimensions).

yup. 

> Feel free to share your thoughts 

All I ever needed is ellipsis and explicit notation (and I like using spaces anywhere, e.g. `'b ... c,  c d -> b ... d '`, but that's just a taste).
Any standard that requires this minimum done is already perfect for me.

I don't use explicit path specification as framework should be in a better position to guess contraction path (that's not just number of operations - but also how memory-friendly operations become, and how much memory available).
Should I need specify a contraction path I'd prefer a number of einsums instead.

Implicit notation is of questionable value to me - I want to see result, not infer it from pattern.
As examples: why `ij,jk` and `cb,bd` are two different operations? Why `a,a` and `...,...` are not the same for 1d tensors? etcetc. 

Index-based interface is useless - when searching github for usages of it, saw dozens of different *CI tests* but I found zero production uses.

**Comment by asmeurer** [comment]
> Index-based interface is useless - when searching github for usages of it, saw dozens of different CI tests but I found zero production uses.

The index interface is more useful for instances where you generate an einsum programmatically.

**Comment by arogozhnikov** [comment]
> The index interface is more useful for instances where you generate an einsum programmatically.

Yes, I guess that was in thinking when designed, but this interleaving of indices and tensors is so awkward - it is not friendly to coding (and specially to type hinting).

**Comment by asmeurer** [comment]
Perhaps the syntax could be improved. In principle an explicit form should be much better for type hinting than a string that is only parsed at runtime.

**Comment by arogozhnikov** [comment]
> In principle an explicit form should be much better for type hinting than a string that is only parsed at runtime.

Why? For type hinting signature `einsum(str, *tensor) -> tensor` is good enough.

To actually check an operation all of them aren't good: `einsum(misplaced_x, x, (1, 2, 4), y, (1, 2, 5, 6, -1), (0, 3, 100, ...))`. 
In strings that's apparent that something is off, but not in indices.

There is some expectation that having tuples makes code more statically checkable, but it is not the case. 
Strings are more human-checkable.


Should there be index interface for einsum (should?), I'd prefer them being separated in API (einops_str, einops_indices or alike).

**Comment by leofang** [comment]
I am very eager to respond in length, but unfortunately I am a bit swamped this week. Nevertheless, the discussion so far is very encouraging, as I was thinking
>> I have a very radical version in mind (some NumPy behavior is simply bad IMHO) 

and I am very happy to know I am not alone
> The index interface is more useful for instances where you generate an einsum programmatically.

> Implicit notation is of questionable value to me - I want to see result, not infer it from pattern.

> I'd prefer them being separated in API (einops_str, einops_indices or alike).

Let me turn on my radical side and add why I dislike the einsum string form (and I promise I'll (try to) be more responsive next week). These are two expressions taken from our test suite:
```
'ÄivrÃEË,JÓIT,ÐOáÙà,Åx,oÒØÏmUÍ,OàW,SgÌCÈ,NgSoJPÌ,vcXz×,HÚGÞ,Æw,ÝD,GsÑi,QÛlÔqzL,IÓqhBdÜ,ÄV,KÎAÇ,uTÖDQÕ,ÆstFÝA,ÛtjÂK,pÔYMbe,FÂkÑÎÁ,rlCVÉÜwaP,ÐÖÀnxU,Þhym,nÏÙßØRÚ,HZÇßk,WÊMË,ÊfÀÍÕLZ,×ÃÈEuRp,ÅjÉáYNB,ÒÁXy->cbefad'
```
and 
```
abc,def,ghi,jkl,mno,pqr,stu,vwx,yzA,BCD,EFG,HIJ,KLM,NOP,QRS,TUV,WXY,ZÀÁ,ÂÃÄ,ÅÆÇ,ÈÉÊ,ËÌÍ,ÎÏÐ,ÑÒÓ,ÔÕÖ,×ØÙ,ÚÛÜ,ÝÞß,àáâ,ãäå,æçè,éêë,cfìí,íîï,ilîð,ðñò,orñó,óôõ,uxôö,ö÷ø,AD÷ù,ùúû,GJúü,üýþ,MPýÿ,ÿĀā,SVĀĂ,ĂăĄ,YÁăą,ąĆć,ÄÇĆĈ,ĈĉĊ,ÊÍĉċ,ċČč,ÐÓČĎ,ĎďĐ,ÖÙďđ,đĒē,ÜßĒĔ,ĔĕĖ,âåĕė,ėĘę,èëĘĚ,Ěìě,ïòĜĝ,ĝĞğ,õøĞĠ,ĠġĢ,ûþġģ,ģĤĥ,āĄĤĦ,ĦħĨ,ćĊħĩ,ĩĪī,čĐĪĬ,ĬĭĮ,ēĖĭį,įİı,ęěİĲ,ĲĜĳ,ğĢĴĵ,ĵĶķ,ĥĨĶĸ,ĸĹĺ,īĮĹĻ,ĻļĽ,ıĳļľ,ľĴĿ,ķĺŀŁ,ŁłŃ,ĽĿłń,ŅņŇ,ňŉŊ,ŋŌō,ŎŏŐ,őŒœ,ŔŕŖ,ŗŘř,ŚśŜ,ŝŞş,ŠšŢ,ţŤť,ŦŧŨ,ũŪū,ŬŭŮ,ůŰű,ŲųŴ,ŵŶŷ,ŸŹź,ŻżŽ,žſƀ,ƁƂƃ,ƄƅƆ,ƇƈƉ,ƊƋƌ,ƍƎƏ,ƐƑƒ,ƓƔƕ,ƖƗƘ,ƙƚƛ,ƜƝƞ,ƟƠơ,ƢƣƤ,ŇŊƥƦ,ƦƧƨ,ōŐƧƩ,Ʃƪƫ,œŖƪƬ,ƬƭƮ,řŜƭƯ,ƯưƱ,şŢưƲ,ƲƳƴ,ťŨƳƵ,ƵƶƷ,ūŮƶƸ,Ƹƹƺ,űŴƹƻ,ƻƼƽ,ŷźƼƾ,ƾƿǀ,Žƀƿǁ,ǁǂǃ,ƃƆǂǄ,Ǆǅǆ,ƉƌǅǇ,Ǉǈǉ,ƏƒǈǊ,Ǌǋǌ,ƕƘǋǍ,ǍǎǏ,ƛƞǎǐ,ǐǑǒ,ơƤǑǓ,Ǔƥǔ,ƨƫǕǖ,ǖǗǘ,ƮƱǗǙ,ǙǚǛ,ƴƷǚǜ,ǜǝǞ,ƺƽǝǟ,ǟǠǡ,ǀǃǠǢ,ǢǣǤ,ǆǉǣǥ,ǥǦǧ,ǌǏǦǨ,ǨǩǪ,ǒǔǩǫ,ǫǕǬ,ǘǛǭǮ,Ǯǯǰ,ǞǡǯǱ,Ǳǲǳ,ǤǧǲǴ,ǴǵǶ,ǪǬǵǷ,ǷǭǸ,ǰǳǹǺ,ǺǻǼ,ǶǸǻǽ,ǽǹǼ,ńŀŃ,çéƠƢ,äæƝƟ,áãƚƜ,ÞàƗƙ,ÛÝƔƖ,ØÚƑƓ,Õ×ƎƐ,ÒÔƋƍ,ÏÑƈƊ,ÌÎƅƇ,ÉËƂƄ,ÆÈſƁ,ÃÅżž,ÀÂŹŻ,XZŶŸ,UWųŵ,RTŰŲ,OQŭů,LNŪŬ,IKŧũ,FHŤŦ,CEšţ,zBŞŠ,wyśŝ,tvŘŚ,qsŕŗ,npŒŔ,kmŏő,hjŌŎ,egŉŋ,bdņň,êaƣŅ->
```
For large tensor networks, it is really an unpleasant interface to work with, we've see expressions that can take a whole slide. There're not many unicode characters that can be used, NumPy/CuPy do not support unicode input, and parsing/processing this kind of long expressions is very tedious. When interfacing a C library like cuTensorNet (or even NumPy internal), the index form is a much better choice.

**Comment by arogozhnikov** [comment]
> happy to know I am not alone

😉

We're talking about very different scenarios - I'm trying to cover common use-case: 2 or at most 4 tensors, usually a direct interface for applied scientist. 

> For large tensor networks, it is really an unpleasant interface to work with

(your argument isn't good - long array of tuples isn't shorter or more readable, API should not target CI as a main scenario). 
Regardless, I agree if you generate TNs from some other abstraction - strings are not an appropriate interface. 

I assume you want to model q entanglement, so I see why you want many indices. <br />
Side remark: einops allows multi-char names (and names can have digits).

Also question to you: aren't you constrained by "single ellipsis"? Do you use ellipsis at all?

**Comment by NeilGirdhar** [comment]
Just in case this is revisited, I wanted to suggest supporting multi-lettered names as per [einops](https://einops.rocks/).

**Comment by arogozhnikov** [comment]
FYI absence of einsum already strikes back:
https://github.com/arogozhnikov/einops/pull/376

Users want to use einsum from einops (mlx does support einsum), but it requires einsum to be exposed in array api.
So users (expectably and reasonably) push to add non-array-api backends. 

MLX is currently the only popular backend where 'default user behavior' is using array API.

**Comment by NeilGirdhar** [comment]
Sorry, but I don't understand what you're saying.  Why does einsum need to be part of the standard?  Can't that operation be implemented using simpler primitives (indexing, multiplication, and addition)?  Why not just make it so that opt-einsum can more easily query an _Array API namesapce_ to figure out whatever it needs to, and then just point to that project?

**Comment by NeilGirdhar** [comment]
> but it is not possible to efficiently implement einsum without knowing a priori sizes and tensor layout (i.e. strides).

Interesting.  Why not expose the tensor layout then?

> I will not directly rely on opt-einsum or similar approach since operation should be at least traceable but better scriptable and compile-able (with support of new sizes of course).

Sorry, but I don't understand this point.

**Comment by arogozhnikov** [comment]
> Sorry, but I don't understand this point.

see https://docs.pytorch.org/docs/stable/generated/torch.jit.trace.html and similar. Tracing / symbolic tracing is a minimalistic technique to collect computation graph. Other techniques are more involved and capture parts of control graph. In jax-land these tools don't exist because of static sizes (which trades it for other problems).



> It seems like every backend will end up redoing this work?

1. frameworks already have einsum, and for reason
2. exact mechanics how it is implemented depends on framework and may/may not include fusion with operations before/after.

As I understand your suggestion:

- let's expose strides => some frameworks won't expose it, specially anything symbolic
- let's build common functions atop exposed APIs  => being traced, they won't be efficient on new shapes

Einsum sits at the right level of abstraction for user-scientist, and provides framework with sufficient flexibility to optimize computations, so yes, it is a good user API and a good framework API, and there is no need to be afraid of it, or try to sweep it under the rug.

Einsum got wide adoption despite being not-like-the-other-function-and-could-be-implemented-by-other-functions-and-appeared-decades-later - proof that worth 10000 posts.

**Comment by arogozhnikov** [comment]
jax traces only static shapes - traced graph won't be applied to a different shape, it is jax's design choice.

**Comment by NeilGirdhar** [comment]
Yes, I know about the static shapes.  I was just curious if you knew whether opt-einsum would do anything that blocks or distorts cotangent propagation?

But regarding your reply, I see your point that if opt-einsum is opaque to Pytorch tracing, then that's a problem.  I wonder if it can be made to work for that case despite non-static shapes.

**Comment by arogozhnikov** [comment]
>  opt-einsum would do anything that blocks or distorts cotangent propagation?

backward in jax+opt-einsum should 'just work'. 

But likely backward implementation is not most efficient in this case, this is something worth checking

**Comment by kgryte** [comment]
@arogozhnikov Given your experience with `einops` and needs articulated here and in https://github.com/data-apis/array-api/issues/494, would you be open to leading the effort to standardize `einsum` in the standard?

You are also welcome to attend the workgroup meetings if you have particular experiences and/or opinions that you'd like to share: https://github.com/data-apis/array-api/issues/853
